# FINAL REPORT ON THE COURSERA CAPSTONE PROJECT

## I- INTRODUCTION / BUSINESS PROBLEM

## II- DATA

This section presents the data I will use to solve the problem. As said in the introduction, I assume the city is the clermont-ferrand town and I am looking for the best place for a new restaurant.

Factors that will influence our decission are:
* number of existing restaurants in the neighborhood (any type of restaurant)
* distance of neighborhood from city center

We first choose to get geographical coordinates of Clermont-Ferrand city, second generate some candidates locations nearby the city center, and third choose to explore some data with the Foursquare API.

### Definition of Foursquare credentials and version

In [1]:
CLIENT_ID = '1TCFP5PIX2MR0LG4BVYIJLL1PGV0Y3Y01IGK5DTHFIAFBF4F' # your Foursquare ID
CLIENT_SECRET = 'ASWAFPQNGFUO0C2G0PAQSKLJMRR5QMNTOHY3S0CFWE24SWEH' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 300
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 1TCFP5PIX2MR0LG4BVYIJLL1PGV0Y3Y01IGK5DTHFIAFBF4F
CLIENT_SECRET:ASWAFPQNGFUO0C2G0PAQSKLJMRR5QMNTOHY3S0CFWE24SWEH


### Import necessary librairies

In [2]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.49-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

### City (Clermont-Ferrand) search for latitude and longitude coordinates

In [3]:
address = ' Jaude, Clermont-Ferrand, CF'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
#location_clermont = geolocator.geocode(address)
clermont_latitude = location.latitude
clermont_longitude = location.longitude
#print(location_clermont[1], location_clermont[0])
print(clermont_latitude, clermont_longitude)

45.7774282 3.0813291


### Definding an url 

In [65]:
#search_query = 'french'
radius = 12000
#url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, clermont_latitude, clermont_longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=1TCFP5PIX2MR0LG4BVYIJLL1PGV0Y3Y01IGK5DTHFIAFBF4F&client_secret=ASWAFPQNGFUO0C2G0PAQSKLJMRR5QMNTOHY3S0CFWE24SWEH&ll=45.7774282,3.0813291&v=20180604&radius=12000&limit=300'

### Send the GET Request and examine the results

In [7]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d31fb702b274a002c76c7f5'},
 'response': {'venues': [{'id': '4b1fedf8f964a520f32a24e3',
    'name': 'Place de Jaude',
    'location': {'address': 'Place de Jaude',
     'lat': 45.77681306296083,
     'lng': 3.0820930571793426,
     'labeledLatLngs': [{'label': 'display',
       'lat': 45.77681306296083,
       'lng': 3.0820930571793426}],
     'distance': 90,
     'postalCode': '63000',
     'cc': 'FR',
     'city': 'Clermont-Ferrand',
     'state': 'Auvergne-Rhône-Alpes',
     'country': 'France',
     'formattedAddress': ['Place de Jaude',
      '63000 Clermont-Ferrand',
      'France']},
    'categories': [{'id': '4bf58dd8d48988d164941735',
      'name': 'Plaza',
      'pluralName': 'Plazas',
      'shortName': 'Plaza',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/plaza_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-1563556720',
    'hasPerk': False},
   {'id': '4c7bf69776ce9c741bddbb

### Get relevant part of JSON and transform it into a *pandas* dataframe

In [8]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
#dataframe = json_normalize(venues)
dataframe = json_normalize(venues)
dataframe

categories  hasPerk  \
0   [{'id': '4bf58dd8d48988d164941735', 'name': 'P...    False   
1   [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...    False   
2   [{'id': '52e81612bcbc57f1066b79f1', 'name': 'B...    False   
3   [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...    False   
4   [{'id': '4bf58dd8d48988d16e941735', 'name': 'F...    False   
5   [{'id': '4bf58dd8d48988d107951735', 'name': 'S...    False   
6   [{'id': '4bf58dd8d48988d132941735', 'name': 'C...    False   
7   [{'id': '52f2ab2ebcbc57f1066b8b4f', 'name': 'B...    False   
8   [{'id': '4bf58dd8d48988d116941735', 'name': 'B...    False   
9   [{'id': '52f2ab2ebcbc57f1066b8b51', 'name': 'T...    False   
10  [{'id': '4bf58dd8d48988d16e941735', 'name': 'F...    False   
11  [{'id': '4bf58dd8d48988d119951735', 'name': 'W...    False   
12  [{'id': '4bf58dd8d48988d136941735', 'name': 'O...    False   
13  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'H...    False   
14  [{'id': '4bf58dd8d48988d123941735', 'name': 'W...    False   
15  [{'id': '4f04afc02fb6e1c99f3db0bc', 'name': 'M...    False   
16  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...    False   
17  [{'id': '4bf58dd8d48988d10c941735', 'name': 'F...    False   
18  [{'id': '4bf58dd8d48988d16c941735', 'name': 'B...    False   
19  [{'id': '4bf58dd8d48988d10a951735', 'name': 'B...    False   
20  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...    False   
21  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...    False   
22  [{'id': '4bf58dd8d48988d1fe941735', 'name': 'M...    False   
23  [{'id': '4bf58dd8d48988d108951735', 'name': 'W...    False   
24  [{'id': '4bf58dd8d48988d116941735', 'name': 'B...    False   
25  [{'id': '4bf58dd8d48988d132941735', 'name': 'C...    False   
26                                                 []    False   
27  [{'id': '4bf58dd8d48988d10a951735', 'name': 'B...    False   
28  [{'id': '4bf58dd8d48988d116941735', 'name': 'B...    False   
29  [{'id': '4bf58dd8d48988d104951735', 'name': 'B...    False   
..                                                ...      ...   
50  [{'id': '4bf58dd8d48988d116941735', 'name': 'B...    False   
51  [{'id': '4bf58dd8d48988d109951735', 'name': 'L...    False   
52  [{'id': '4bf58dd8d48988d1c5941735', 'name': 'S...    False   
53  [{'id': '4bf58dd8d48988d155941735', 'name': 'G...    False   
54  [{'id': '4bf58dd8d48988d11a941735', 'name': 'O...    False   
55  [{'id': '4bf58dd8d48988d123951735', 'name': 'S...    False   
56  [{'id': '4bf58dd8d48988d171941735', 'name': 'E...    False   
57  [{'id': '4bf58dd8d48988d1fb941735', 'name': 'H...    False   
58  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...    False   
59  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...    False   
60  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...    False   
61  [{'id': '4d954afda243a5684865b473', 'name': 'O...    False   
62  [{'id': '5032885091d4c4b30a586d66', 'name': 'R...    False   
63  [{'id': '4bf58dd8d48988d104951735', 'name': 'B...    False   
64  [{'id': '4bf58dd8d48988d114951735', 'name': 'B...    False   
65  [{'id': '4bf58dd8d48988d109951735', 'name': 'L...    False   
66  [{'id': '4bf58dd8d48988d130941735', 'name': 'B...    False   
67  [{'id': '4bf58dd8d48988d130941735', 'name': 'B...    False   
68  [{'id': '4bf58dd8d48988d1f6931735', 'name': 'G...    False   
69  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'H...    False   
70  [{'id': '4bf58dd8d48988d111951735', 'name': 'J...    False   
71  [{'id': '4d954b06a243a5684965b473', 'name': 'R...    False   
72  [{'id': '4bf58dd8d48988d10a951735', 'name': 'B...    False   
73  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...    False   
74  [{'id': '4bf58dd8d48988d125941735', 'name': 'T...    False   
75  [{'id': '52f2ab2ebcbc57f1066b8b27', 'name': 'M...    False   
76  [{'id': '4bf58dd8d48988d124941735', 'name': 'O...    False   
77  [{'id': '4bf58dd8d48988d1f8941735', 'name': 'F...    False   
78  [{'id': '4bf58dd8d48988d1e5931735', 'name': 'M...    False   
79  [{'id': '4bf5

### Define information of interest and filter dataframe

In [9]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name  \
0                        Place de Jaude   
1                        Le Faisan Doré   
2                   Le Bistrot d'à Côté   
3                   Le Bouchon de Jaude   
4                            McDonald's   
5                               Minelli   
6       Église Saint-Pierre-des-Minimes   
7                 Arrêt Jaude ⒷⒸ [9,10]   
8                     Brasserie Le Lion   
9                      Station Jaude  Ⓐ   
10                                Quick   
11                   La Cave Du Théâtre   
12  Opéra Municipal de Clermont-Ferrand   
13                        Hotel Le Lion   
14                          La Régalade   
15                      Boutique Orange   
16                       La Gourmandine   
17                   Lard et la Manière   
18                          Big Fernand   
19                   Agence BNP Paribas   
20                              Filprod   
21                   Le Fournil De Jean   
22                     Clermont musique   
23                      marche du temps   
24                        Hotel De Lyon   
25                St Pierre les Minimes   
26                                Mango   
27                     Société Générale   
28                          The Embassy   
29                              bizzbee   
..                                  ...   
50                           Le Voilier   
51                      Boutique aubade   
52                   Le Fournil Du Papé   
53                            Au Bureau   
54                       Le Saison Doré   
55                       Tabac Du globe   
56                Espace Pierre Laporte   
57                              Aniline   
58                          Quicksilver   
59                           Le Débilum   
60                          Epi Gaulois   
61                        Jaude Optique   
62                         Régie Mialon   
63                                Magic   
64                              Book'In   
65                    Danielle Boutique   
66                       Adrea Mutuelle   
67                        Mission Local   
68                 Nouvelles Frontieres   
69       Citea Residentiel Gerzat Hôtel   
70                               Agatha   
71                         Courty Tbite   
72                    LCL Banque Privée   
73                                    Z   
74                                  DBM   
75                          Carré Blanc   
76                             Barclays   
77                    Envie d'intérieur   
78         Auditorium Emmanuel Chabrier   
79                         Tabac Presse   

                                  categories  \
0                                      Plaza   
1                          French Restaurant   
2                                     Bistro   
3                          French Restaurant   
4                       Fast Food Restaurant   
5                                 Shoe Store   
6                                     Church   
7                                   Bus Stop   
8                                        Bar   
9                               Tram Station   
10                      Fast Food Restaurant   
11                                 Wine Shop   
12                               Opera House   
13                                     Hotel   
14                                  Wine Bar   
15                         Mobile Phone Shop   
16                         French Restaurant   
17                         French Restaurant   
18                              Burger Joint   
19                                      Bank   
20                     General Entertainment   
21                                    Bakery   
22                               Music Store   
23                             Women's Store   
24                                       Bar   
25                                    Church   
26                                      None   
27                      

In [134]:
#dataframe_filtered['restaurant' in dataframe_filtered['categories']]

# III- METHODOLOGY 

## Description

## Analysis

### Converting Code

This codes with helps to convert coordinates data

In [11]:
!pip install shapely
import shapely.geometry

!pip install pyproj
import pyproj

import math

def lonlat_to_xy(lon, lat):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    xy = pyproj.transform(proj_latlon, proj_xy, lon, lat)
    return xy[0], xy[1]

def xy_to_lonlat(x, y):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    lonlat = pyproj.transform(proj_xy, proj_latlon, x, y)
    return lonlat[0], lonlat[1]

def calc_xy_distance(x1, y1, x2, y2):
    dx = x2 - x1
    dy = y2 - y1
    return math.sqrt(dx*dx + dy*dy)

### Creation of Hexagonal grid
Let's create a hexagonal grid of cells: we offset every other row, and adjust vertical row spacing so that every cell center is equally distant from all it's neighbors

In [12]:
clermont_center_x, clermont_center_y = lonlat_to_xy(clermont_longitude, clermont_latitude) # City center in Cartesian coordinates

k = math.sqrt(3) / 2 # Vertical offset for hexagonal grid cells
x_min = clermont_center_x - 6000
x_step = 600
y_min = clermont_center_y - 6000 - (int(21/k)*k*600 - 12000)/2
y_step = 600 * k 

latitudes = []
longitudes = []
distances_from_center = []
xs = []
ys = []
for i in range(0, int(21/k)):
    y = y_min + i * y_step
    x_offset = 300 if i%2==0 else 0
    for j in range(0, 21):
        x = x_min + j * x_step + x_offset
        distance_from_center = calc_xy_distance(clermont_center_x, clermont_center_y, x, y)
        if (distance_from_center <= 6001):
            lon, lat = xy_to_lonlat(x, y)
            latitudes.append(lat)
            longitudes.append(lon)
            distances_from_center.append(distance_from_center)
            xs.append(x)
            ys.append(y)
#print(latitudes)
#print(len(latitudes), 'candidate neighborhood centers generated.')

In [13]:
df_locations = pd.DataFrame({#'Address': addresses,
                             'Latitude': latitudes,
                             'Longitude': longitudes,
                             'X': xs,
                             'Y': ys,
                             'Distance from center': distances_from_center})

df_locations.head(10)

Latitude  Longitude              X             Y  Distance from center
0  45.724696   3.069564 -428167.930747  5.133149e+06           5992.495307
1  45.725495   3.077108 -427567.930747  5.133149e+06           5840.376700
2  45.726294   3.084652 -426967.930747  5.133149e+06           5747.173218
3  45.727092   3.092196 -426367.930747  5.133149e+06           5715.767665
4  45.727890   3.099741 -425767.930747  5.133149e+06           5747.173218
5  45.728687   3.107286 -425167.930747  5.133149e+06           5840.376700
6  45.729483   3.114832 -424567.930747  5.133149e+06           5992.495307
7  45.728072   3.057259 -429067.930747  5.133669e+06           5855.766389
8  45.728872   3.064803 -428467.930747  5.133669e+06           5604.462508
9  45.729671   3.072348 -427867.930747  5.133669e+06           5408.326913

In [14]:
df_locations.to_pickle('./locations.pkl')  

In [16]:
map_clermont = folium.Map(location=[clermont_latitude, clermont_longitude], zoom_start=13)
folium.Marker([clermont_latitude, clermont_longitude], popup='Jaude, Clermont-Ferrand').add_to(map_clermont)
for lat, lon in zip(latitudes, longitudes):
    #folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_berlin) 
    folium.Circle([lat, lon], radius=300, color='blue', fill=False).add_to(map_clermont)
    #folium.Marker([lat, lon]).add_to(map_berlin)
map_clermont

In [17]:
#food_category = '4d4b7105d754a06374d81259' # 'Root' category for all food-related venues
def is_restaurant(categories, specific_filter=None):
    restaurant_words = ['restaurant', 'diner', 'taverna', 'steakhouse']
    restaurant = False
    specific = False
    for c in categories:
        category_name = c[0].lower()
        category_id = c[1]
        for r in restaurant_words:
            if r in category_name:
                restaurant = True
        if 'fast food' in category_name:
            restaurant = False
        if not(specific_filter is None) and (category_id in specific_filter):
            specific = True
            restaurant = True
    return restaurant, specific

def get_categories(categories):
    return [(cat['name'], cat['id']) for cat in categories]

def format_address(location):
    address = ', '.join(location['formattedAddress'])
    return address

def get_venues_near_location(lat, lon, category, client_id, client_secret, radius=500, limit=100):
    version = '20180724'
    url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&categoryId={}&radius={}&limit={}'.format(
        client_id, client_secret, version, lat, lon, category, radius, limit)
    try:
        results = requests.get(url).json()['response']['groups'][0]['items']
        venues = [(item['venue']['id'],
                   item['venue']['name'],
                   get_categories(item['venue']['categories']),
                   (item['venue']['location']['lat'], item['venue']['location']['lng']),
                   format_address(item['venue']['location']),
                   item['venue']['location']['distance']) for item in results]        
    except:
        venues = []
    return venues

In [18]:
# Let's now go over our neighborhood locations and get nearby restaurants; we'll also maintain a dictionary of all found restaurants and all found italian restaurants

import pickle

def get_restaurants(lats, lons):
    restaurants = {}
    african_restaurants = {}
    location_restaurants = []

    print('Obtaining venues around candidate locations:', end='')
    for lat, lon in zip(lats, lons):
        # Using radius=350 to meke sure we have overlaps/full coverage so we don't miss any restaurant (we're using dictionaries to remove any duplicates resulting from area overlaps)
        venues = get_venues_near_location(lat, lon, food_category, CLIENT_ID, CLIENT_SECRET, radius=350, limit=100)
        area_restaurants = []
        for venue in venues:
            venue_id = venue[0]
            venue_name = venue[1]
            venue_categories = venue[2]
            venue_latlon = venue[3]
            venue_address = venue[4]
            venue_distance = venue[5]
            #is_res, is_african = is_restaurant(venue_categories, specific_filter=italian_restaurant_categories)
            is_res, is_african = is_restaurant(venue_categories)
            if is_res:
                x, y = lonlat_to_xy(venue_latlon[1], venue_latlon[0])
                restaurant = (venue_id, venue_name, venue_latlon[0], venue_latlon[1], venue_address, venue_distance, is_african, x, y)
                if venue_distance<=300:
                    area_restaurants.append(restaurant)
                restaurants[venue_id] = restaurant
                if is_african:
                    african_restaurants[venue_id] = restaurant
        location_restaurants.append(area_restaurants)
        print(' .', end='')
    print(' done.')
    return restaurants, african_restaurants, location_restaurants

# Try to load from local file system in case we did this before
restaurants = {}
african_restaurants = {}
location_restaurants = []
loaded = False

restaurants, african_restaurants, location_restaurants = get_restaurants(latitudes, longitudes)

# Let's persists this in local file system
with open('restaurants_350.pkl', 'wb') as f:
    pickle.dump(restaurants, f)
with open('african_restaurants_350.pkl', 'wb') as f:
    pickle.dump(african_restaurants, f)
with open('location_restaurants_350.pkl', 'wb') as f:
    pickle.dump(location_restaurants, f)


Obtaining venues around candidate locations: . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . done.


In [19]:
import numpy as np

print('Total number of restaurants:', len(restaurants))
print('Total number of african restaurants:', len(african_restaurants))
print('Percentage of african restaurants: {:.2f}%'.format(len(african_restaurants) / len(restaurants) * 100))
print('Average number of restaurants in neighborhood:', np.array([len(r) for r in location_restaurants]).mean())

Total number of restaurants: 152
Total number of african restaurants: 0
Percentage of african restaurants: 0.00%
Average number of restaurants in neighborhood: 0.3543956043956044


In [20]:
print('List of all restaurants')
print('-----------------------')
for r in list(restaurants.values())[:10]:
    print(r)
print('...')
print('Total:', len(restaurants))

List of all restaurants
-----------------------
('584003a16f20d11b223c07ad', 'CBH Cuisines Professionnelles', 45.74258002149345, 3.1018352508544917, '13 bis rue Cugnot, 63540 ROMAGNAT, France', 117, False, -425359.0246018602, 5134755.757146825)
('5270fa6111d2c97eeaa622cc', 'Rouge Tendance', 45.75312997180263, 3.140796381479397, 'France', 241, False, -422154.3514284204, 5135470.703306506)
('4f390034e4b0948a83a52600', 'Cafeteria Auchan', 45.752402506847694, 3.128564517897922, 'France', 218, False, -423117.1267175402, 5135533.023429573)
('5494146c498ec92862e2d1ea', '5ème Saison', 45.75284957885742, 3.136699914932251, '34 Avenue De COURNON, 63170 Aubiere, France', 257, False, -422477.3976580292, 5135487.476067202)
('4e674d62ae609d64bcd17c1c', 'La Distillerie', 45.75727138796016, 3.1330977411735095, '39 rue des Sauzes, 63170 Aubière, France', 258, False, -422683.44720677566, 5136020.504812379)
('4ca8c8668882199c6fc744d4', 'Oïshi', 45.756805078836095, 3.1455516915886217, '49 avenue Lavoisier

In [21]:
print('Restaurants around location')
print('---------------------------')
for i in range(100, 110):
    rs = location_restaurants[i][:8]
    names = ', '.join([r[1] for r in rs])
    print('Restaurants around location {}: {}'.format(i+1, names))

Restaurants around location
---------------------------
Restaurants around location 101: 
Restaurants around location 102: 
Restaurants around location 103: 
Restaurants around location 104: 
Restaurants around location 105: Cantine des Impots, Le Petit Bonneval, Piz'n'Wok
Restaurants around location 106: Le Viaduc
Restaurants around location 107: 
Restaurants around location 108: 
Restaurants around location 109: 
Restaurants around location 110: kanka family kebab, Kanka Restaurant


### Visualization of restaurants in the vicinity of the city center

In [23]:
map_clermont = folium.Map(location=[clermont_latitude, clermont_longitude], zoom_start=13)
folium.Marker([clermont_latitude, clermont_longitude], popup='Jaude, Clermont-Ferrand, France').add_to(map_clermont)
for res in restaurants.values():
    lat = res[2]; lon = res[3]
    is_african = res[6]
    color = 'red'
    folium.CircleMarker([lat, lon], radius=3, color=color, fill=True, fill_color=color, fill_opacity=1).add_to(map_clermont)
map_clermont

### Visualization of restaurants and locations 

Now that we have the distribution of the restaurants in the Clermont-city in a radiums of 6km from the city center jaude, we can visualize them with the locations

In [32]:
map_clermont = folium.Map(location=[clermont_latitude, clermont_longitude], zoom_start=13)
folium.Marker([clermont_latitude, clermont_longitude], popup='Jaude, Clermont-Ferrand, France').add_to(map_clermont)
for res in restaurants.values():
    lat = res[2]; lon = res[3]
    is_african = res[6]
    color = 'red'
    folium.CircleMarker([lat, lon], radius=3, color=color, fill=True, fill_color=color, fill_opacity=1).add_to(map_clermont)
for lat, lon in zip(latitudes, longitudes):
    folium.CircleMarker([lat, lon], radius=3, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_clermont)
map_clermont

### Visualization of restaurants and locations using HeatMap

In [25]:
map_clermontfrom folium.plugins import HeatMap

In [39]:
restaurant_latlons = [[res[2], res[3]] for res in restaurants.values()]

map_clermont = folium.Map(location=[clermont_latitude, clermont_longitude], zoom_start=13, tiles='Stamen Terrain')
folium.Marker([clermont_latitude, clermont_longitude], popup='Jaude, Clermont-Ferrand, France').add_to(map_clermont)
folium.TileLayer('cartodbpositron').add_to(map_clermont)
HeatMap(restaurant_latlons).add_to(map_clermont)
for lat, lon in zip(latitudes, longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_clermont)
map_clermont

### Couting of locations restaurants  

In [40]:
location_restaurants_count = [len(res) for res in location_restaurants]

df_locations['Restaurants in area'] = location_restaurants_count

print('Average number of restaurants in every area with radius=300m:', np.array(location_restaurants_count).mean())

df_locations

Average number of restaurants in every area with radius=300m: 0.3543956043956044


Latitude  Longitude              X             Y  Distance from center  \
0    45.724696   3.069564 -428167.930747  5.133149e+06           5992.495307   
1    45.725495   3.077108 -427567.930747  5.133149e+06           5840.376700   
2    45.726294   3.084652 -426967.930747  5.133149e+06           5747.173218   
3    45.727092   3.092196 -426367.930747  5.133149e+06           5715.767665   
4    45.727890   3.099741 -425767.930747  5.133149e+06           5747.173218   
5    45.728687   3.107286 -425167.930747  5.133149e+06           5840.376700   
6    45.729483   3.114832 -424567.930747  5.133149e+06           5992.495307   
7    45.728072   3.057259 -429067.930747  5.133669e+06           5855.766389   
8    45.728872   3.064803 -428467.930747  5.133669e+06           5604.462508   
9    45.729671   3.072348 -427867.930747  5.133669e+06           5408.326913   
10   45.730470   3.079892 -427267.930747  5.133669e+06           5273.518749   
11   45.731269   3.087437 -426667.930747  5.133669e+06           5204.805472   
12   45.732067   3.094982 -426067.930747  5.133669e+06           5204.805472   
13   45.732865   3.102528 -425467.930747  5.133669e+06           5273.518749   
14   45.733662   3.110074 -424867.930747  5.133669e+06           5408.326913   
15   45.734458   3.117620 -424267.930747  5.133669e+06           5604.462508   
16   45.735254   3.125167 -423667.930747  5.133669e+06           5855.766389   
17   45.731446   3.044953 -429967.930747  5.134189e+06           5901.694672   
18   45.732247   3.052497 -429367.930747  5.134189e+06           5556.077753   
19   45.733048   3.060042 -428767.930747  5.134189e+06           5256.424640   
20   45.733847   3.067587 -428167.930747  5.134189e+06           5010.987927   
21   45.734647   3.075132 -427567.930747  5.134189e+06           4828.043082   
22   45.735446   3.082677 -426967.930747  5.134189e+06           4714.870094   
23   45.736244   3.090223 -426367.930747  5.134189e+06           4676.537180   
24   45.737042   3.097769 -425767.930747  5.134189e+06           4714.870094   
25   45.737839   3.105315 -425167.930747  5.134189e+06           4828.043082   
26   45.738636   3.112862 -424567.930747  5.134189e+06           5010.987927   
27   45.739433   3.120409 -423967.930747  5.134189e+06           5256.424640   
28   45.740229   3.127956 -423367.930747  5.134189e+06           5556.077753   
29   45.741024   3.135504 -422767.930747  5.134189e+06           5901.694672   
..         ...        ...            ...           ...                   ...   
334  45.813799   3.027083 -429967.930747  5.143542e+06           5901.694672   
335  45.814602   3.034638 -429367.930747  5.143542e+06           5556.077753   
336  45.815405   3.042193 -428767.930747  5.143542e+06           5256.424640   
337  45.816207   3.049748 -428167.930747  5.143542e+06           5010.987927   
338  45.817009   3.057304 -427567.930747  5.143542e+06           4828.043082   
339  45.817810   3.064860 -426967.930747  5.143542e+06           4714.870094   
340  45.818611   3.072416 -426367.930747  5.143542e+06           4676.537180   
341  45.819411   3.079973 -425767.930747  5.143542e+06           4714.870094   
342  45.820211   3.087530 -425167.930747  5.143542e+06           4828.043082   
343  45.821010   3.095087 -424567.930747  5.143542e+06           5010.987927   
344  45.821808   3.102645 -423967.930747  5.143542e+06           5256.424640   
345  45.822607   3.110203 -423367.930747  5.143542e+06           5556.077753   
346  45.823404   3.117761 -422767.930747  5.143542e+06           5901.694672   
347  45.819579   3.037421 -429067.930747  5.144061e+06           5855.766389   
348  45.820381   3.044977 -428467.930747  5.144061e+06           5604.462508   
349  45.821183   3.052533 -427867.930747  5.144061e+06           5408.326913   
350  45.821985   3.060089 -427267.930747  5.144061e+06           5273.518749   
351  45.822786   3.067646 -426667.930747  5.144061e+06           5204.805472   
352  4

In [42]:
df_locations_max = df_locations[df_locations['Restaurants in area']>2]
df_locations_max

Latitude  Longitude              X             Y  Distance from center  \
75   45.755948   3.143876 -421867.930747  5.135748e+06           5474.486277   
93   45.760127   3.139118 -422167.930747  5.136267e+06           4938.623290   
104  45.758725   3.081510 -426667.930747  5.136787e+06           2100.000000   
139  45.765477   3.056884 -428467.930747  5.137826e+06           2343.074903   
162  45.772852   3.082318 -426367.930747  5.138346e+06            519.615242   
167  45.776842   3.120075 -423367.930747  5.138346e+06           3044.667470   
179  45.775428   3.062452 -427867.930747  5.138865e+06           1500.000000   
182  45.777828   3.085105 -426067.930747  5.138865e+06            300.000000   
201  45.782004   3.080340 -426367.930747  5.139385e+06            519.615242   
203  45.783602   3.095444 -425167.930747  5.139385e+06           1307.669683   
204  45.784400   3.102996 -424567.930747  5.139385e+06           1873.499400   
209  45.788384   3.140764 -421567.930747  5.139385e+06           4828.043082   
222  45.787779   3.090679 -425467.930747  5.139904e+06           1374.772708   
331  45.817631   3.107412 -423667.930747  5.143022e+06           4956.813493   

     Restaurants in area  
75                     6  
93                     4  
104                    3  
139                    3  
162                    6  
167                    3  
179                    3  
182                   12  
201                    3  
203                    3  
204                    3  
209                    3  
222                    4  
331                    3

In [43]:
df_locations_min = df_locations[df_locations['Restaurants in area']<=2]
df_locations_min.head()

Latitude  Longitude              X             Y  Distance from center  \
0  45.724696   3.069564 -428167.930747  5.133149e+06           5992.495307   
1  45.725495   3.077108 -427567.930747  5.133149e+06           5840.376700   
2  45.726294   3.084652 -426967.930747  5.133149e+06           5747.173218   
3  45.727092   3.092196 -426367.930747  5.133149e+06           5715.767665   
4  45.727890   3.099741 -425767.930747  5.133149e+06           5747.173218   

   Restaurants in area  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0

### Clustering of the data

We then cluster with the k-means algortihm the areas with areas less or equals to 2 restaurants. We choose to have 4 candidates clusters and plot their centers. K-means algorithm is chose because of its simplicity of implementation.

In [46]:
from sklearn.cluster import KMeans

number_of_clusters = 4

#good_xys = df_good_locations[['X', 'Y']].values
good_xys = df_locations_min[['X', 'Y']].values
restaurant_latlons = [[res[2], res[3]] for res in restaurants.values()]

kmeans = KMeans(init='k-means++', n_clusters=number_of_clusters, random_state=0, n_init=12).fit(good_xys)

cluster_centers = [xy_to_lonlat(cc[0], cc[1]) for cc in kmeans.cluster_centers_]

map_clermont = folium.Map(location=[clermont_latitude, clermont_longitude], zoom_start=13)
folium.TileLayer('cartodbpositron').add_to(map_clermont)
HeatMap(restaurant_latlons).add_to(map_clermont)
for lon, lat in cluster_centers:
    folium.Circle([lat, lon], radius=500, color='green', fill=True, fill_opacity=0.25).add_to(map_clermont) 
for lat, lon in zip(latitudes, longitudes):
    folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_clermont)
map_clermont

In [47]:
cluster_centers

[(3.042817670315043, 45.794928060853266),
 (3.105863098291043, 45.80606883790682),
 (3.119188603231289, 45.759568680689846),
 (3.0560261375214237, 45.74935328906884)]

In [58]:
def get_address(api_key, latitude, longitude, verbose=False):
    try:
        url = 'https://maps.googleapis.com/maps/api/geocode/json?key={}&latlng={},{}&sensor=true'.format(api_key, latitude, longitude)
        #version = '20180724'
        #url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        client_id, client_secret, version, lat, lon, 1, 1)
        response = requests.get(url).json()
        if verbose:
            print('Google Maps API JSON result =>', response)
        results = response['results']
        #address = results[0]['formatted_address']
        print(results)
        return address
    except:
        return None

In [59]:
for lon, lat in cluster_centers:
    addr = get_address('AIzaSyAejjLL7ciZ6of0DiPOAaDAPrlyZsw6phY', lat, lon)
    print(addr)

None
None
None
None


In [61]:
#Installation of geocoder module
! pip install geocoder 

     |████████████████████████████████| 102kB 4.4MB/s ta 0:00:01


In [64]:
import geocoder
for lon, lat in cluster_centers:
    addr = geocoder.google([lat, lon], method='reverse')
    print(addr)

<[REQUEST_DENIED] Google - Reverse [empty]>
<[REQUEST_DENIED] Google - Reverse [empty]>
<[REQUEST_DENIED] Google - Reverse [empty]>
<[REQUEST_DENIED] Google - Reverse [empty]>


# IV- RESULTS

We conduct our analysis for the town of Clermont-Ferrand. We fixed to search locations around near the city centercalled Jaude. In terms of affluence, the city center will have more population and transports.   

The analysis of the data shows that if we limit the research of the area of 6km around the city center, we have many candidates of locations.
When directing our attention on the distribution of restaurants in these locations. The map reveal that the density of restaurants is higher in the west part of the vicinity of the city center, than other part of the ciy center.     

We then decided to look at locations with less or equals number of restaurants in the vicinity to 2. After having the candidates, we clustered them into 4 using the K_means algorithm. The results are 4 clusters centers where the centers are good candidates to be the best place for the new restaurant. The corresponding address of these places are :
* Route de Ternant, 63830 Durtol, France
* Fédération PCF 63, 34 Rue des Clos, 63100 Clermont-Ferrand, France
* 18 Allée du Capitaine Diéderich, 63170 Aubière, France
* 39 Avenue du Chorigier, 63122 Ceyrat, France 

# V- DISCUSSION

# VI- CONCLUSION 